### PR TITLE
add gitlab-ci and helm chart

### PR DESCRIPTION
**Pull request by ArshamTeymouri**

### CODE REVIEW

File: helm/values.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+common:
+  fullnameOverride: "pr-cutom-review"
```
Comment (sergejparity) [inline helm/values.yaml:2]:
```suggestion
  fullnameOverride: "pr-custom-review"
```

File: helm/values-prod.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+common:
+  secrets:
+    GITHUB_ACCESS_TOKEN: ref+vault://kv/cicd/gitlab/parity/opstooling/pr-custom-review/parity-prod#DB_PASSWORD
```
Comment (sergejparity) [inline helm/values-prod.yaml:3]:
```suggestion
    GITHUB_ACCESS_TOKEN: ref+vault://kv/cicd/gitlab/parity/opstooling/pr-custom-review/parity-prod#GITHUB_ACCESS_TOKEN
```

I think it's better to keep names consistent.

File: helm/values.yaml (hunk 1 above)
Comment (ArshamTeymouri) [reply helm/values.yaml:2]: nice catch!

**Review by sergejparity — APPROVED** [review]

File: helm/values.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    httpGet:
+      path: /ping
+    periodSeconds: 10
+  readinessProbe:
+    failureThreshold: 3
+    successThreshold: 1
+    initialDelaySeconds: 0
+    timeoutSeconds: 1
+    httpGet:
+      path: /ping
```
Comment (joao-paulo-parity) [inline helm/values.yaml:36]: `/ping` does not exist yet but I'll push a commit adding it

File: helm/values.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    type: ClusterIP
+    port: 80
+    containerPort: 3000
+  livenessProbe:
+    failureThreshold: 3
+    successThreshold: 1
+    initialDelaySeconds: 0
+    timeoutSeconds: 1
+    httpGet:
+      path: /ping
```
Comment (joao-paulo-parity) [inline helm/values.yaml:28]: `/ping` does not exist yet but I'll push a commit adding it

**Review by joao-paulo-parity — CHANGES_REQUESTED** [review]

File: helm/values.yaml (hunk 3 above)
Comment (joao-paulo-parity) [reply helm/values.yaml:36]: d46b3e80165d1467da0ba3be388dd9a769918132

File: helm/values.yaml (hunk 4 above)
Comment (joao-paulo-parity) [reply helm/values.yaml:28]: d46b3e80165d1467da0ba3be388dd9a769918132

**Review by joao-paulo-parity — APPROVED** [review]

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  image:                           $CI_IMAGE
+
+.common-refs:                      &common-refs
+  rules:
+    - if: $CI_COMMIT_REF_NAME == "master"
+    - if: $CI_COMMIT_REF_NAME =~ /^[0-9]+$/                 # PRs
+
+.staging-refs:                     &staging-refs
+  rules:
+    - if: $CI_COMMIT_REF_NAME =~ /^stg-v[0-9]+\.[0-9]+.*$/  # e.g. stg-v0.1
```
Comment (joao-paulo-parity) [inline .gitlab-ci.yml:38]: `$CI_COMMIT_TAG` would be more precise here but it doesn't make too much of a difference

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+.production-refs:                  &production-refs
+  rules:
+    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/      # e.g. v0.1
```
Comment (joao-paulo-parity) [inline .gitlab-ci.yml:42]: `$CI_COMMIT_TAG` would be more precise here but it doesn't make too much of a difference

Comment (joao-paulo-parity) [reply .gitlab-ci.yml]: a1c5b7a8538bf04bf023f8f7da56c74a377f0520

File: .gitlab-ci.yml (hunk 5 above)
Comment (joao-paulo-parity) [reply .gitlab-ci.yml]: a1c5b7a8538bf04bf023f8f7da56c74a377f0520

**Comment by joao-paulo-parity** [comment]
> Also maybe it make sense to deploy staging on each commit to master

The vast majority of OpsTooling projects have staging deployment on `stg-v*` or `pre-v*` tags. That's what we're used to do and so I think this PR should follow suit. Auto deployment from master isn't something we're interested in having currently as corroborated by https://github.com/paritytech/substrate-matrix-faucet/pull/146 but it could be suggested as a **new** practice within the team. Personally I'm not interested in that and would prefer to stay with tags.

**Review by sergejparity — DISMISSED** [review]

**Review by sergejparity — APPROVED** [review]

File: helm/values.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  podSecurityContext:
+    fsGroup: 1000
+    runAsNonRoot: true
+    runAsUser: 1000
+    runAsGroup: 1000
+  env:
+    LOG_FORMAT: json
+    PORT: 3000
+  secrets:
+    GITHUB_ACCESS_TOKEN: ref+vault://kv/cicd/gitlab/parity/mirrors/pr-custom-review#GITHUB_ACCESS_TOKEN
```
Comment (joao-paulo-parity) [inline helm/values.yaml:16]: Tried this PR

https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1578420#L45

complains about something related to base64

**Review by joao-paulo-parity — CHANGES_REQUESTED** [review]

File: helm/values.yaml (hunk 7 above)
Comment (sergejparity) [reply helm/values.yaml:16]: Well probably we hit this issue: https://stackoverflow.com/questions/53394973/cant-create-secret-in-kubernetes-illegal-base64-data-at-input

Comment (ArshamTeymouri) [reply helm/values.yaml:16]: just base64 encoded the secret, please try one more time.

Comment (joao-paulo-parity) [reply helm/values.yaml:16]: > just base64 encoded the secret, please try one more time.

Didn't work https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1578629#L44

File: helm/values.yaml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    create: false
+  podSecurityContext:
+    fsGroup: 1000
+    runAsNonRoot: true
+    runAsUser: 1000
+    runAsGroup: 1000
+  env:
+    LOG_FORMAT: json
+    PORT: 3000
+  secrets:
```
Comment (sergejparity) [inline helm/values.yaml:15]:
```suggestion
  secrets:
```
```suggestion
```

Comment (sergejparity) [inline helm/values.yaml:15]:
```suggestion
  secrets:
```
```suggestion
```

Comment (joao-paulo-parity) [reply helm/values.yaml]: https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1579390 is the job running for this change

Comment (joao-paulo-parity) [reply helm/values.yaml]: https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1579265 timed out and https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1579390 as well

> Error: release pr-custom-review failed, and has been uninstalled due to atomic being set: timed out waiting for the condition

Comment (sergejparity) [reply helm/values.yaml]: Yeah. `secret "pr-custom-review" not found: CreateContainerConfigError` need to cut more code :)

**Review by joao-paulo-parity — APPROVED** [review]
Works according to https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/pipelines/194920

File: .gitlab-ci.yml [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  script:
+    # TODO: Use pre-commit run --all
+    - echo "lint"
+
+security-scan:
+  <<:                              *common-refs
+  stage:                           test
+  image:                           aquasec/trivy:latest
+  script:
+    - trivy fs .
```
Comment (joao-paulo-parity) [inline .gitlab-ci.yml:68]: security-scan is failing with `503 Service Unavailable` https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1581967#L33

I think the job is implemented correctly, so no code changes needed here.

I think we can disable it and create a ticket to re-enable it after this PR is merged.

Comment (sergejparity) [reply .gitlab-ci.yml:68]: Looks like it was temporary issue. Now it's working https://gitlab.parity.io/parity/mirrors/pr-custom-review/-/jobs/1582086

**Review by sergejparity — APPROVED** [review]

**Pull request merged**